File: daily_problem/2023-3/1616.cpp
/*
1616. 分割两个字符串得到回文串
https://leetcode.cn/problems/split-two-strings-to-make-palindrome/description/
*/

#include <string>

using std::string;

class Solution {
public:
  bool checkPalindromeFormation(string a, string b) {
    return check(a, b) || check(b, a);
  }

  bool check(string &a, string &b) {
    int l = 0, r = a.size() - 1;
    while (l < r && a[l] == b[r]) {
      ++l;
      --r;
    }
    return ifPalindrome(a, l, r) || ifPalindrome(b, l, r);
  }

  bool ifPalindrome(string &a, int l, int r) {
    while (l <= r && a[l] == a[r]) {
      ++l;
      --r;
    }
    return l > r;
  }
};

File: daily_problem/2023-3/1616.py
"""
1616. 分割两个字符串得到回文串
https://leetcode.cn/problems/split-two-strings-to-make-palindrome/description/
"""


class Solution:
    def checkPalindromeFormation(self, a: str, b: str) -> bool:
        n = len(a)

        def check(a, b):
            l, r = 0, n - 1
            while l < r and a[l] == b[r]:
                l += 1
                r -= 1
            s, t = a[l:r+1], b[l:r+1]
            return s == s[::-1] or t == t[::-1]
        return check(a, b) or check(b, a)
